### PR TITLE
[DNM] peer gc raft_log itself, not by leader CompactLog command

### PR DESCRIPTION
**Pull request by javaforfun**

### CODE REVIEW

**Comment by ngaut** [comment]
Does it relate to region checksum? @busyjay

**Comment by javaforfun** [comment]
Seems default [raft_log_gc_count_limit](../blob/master/src/raftstore/store/config.rs#L30) and [raft_log_gc_size_limit](../blob/master/src/raftstore/store/config.rs#L31) are too large for peer gc.  As follower don't know replicated_index, it can only use apply index for compact.

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -970,6 +970,45 @@ impl ApplyDelegate {
...
+            // old format compact log command, safe to ignore.
+            return None;
+        }
+
+        // compact failure is safe to be omitted, no need to assert.
+        if compact_raft_log(&self.tag,
+                            &mut self.apply_state,
+                            compact_index,
+                            compact_term)
+            .is_err() {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: log error here

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1423,15 +1424,13 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-
-            if let Err(e) = self.sendch.try_send(Msg::new_raft_cmd(request, Box::new(|_| {}))) {
-                error!("{} send compact log {} err {:?}", peer.tag, compact_idx, e);
-            }
+            let compact = LocalCompactLog::new(region_id, compact_idx, term);
+            compact_logs.push(compact);
         }
 
+        if !compact_logs.is_empty() {
+            self.apply_worker.schedule(ApplyTask::local_compact_logs(compact_logs)).unwrap();
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: why not sending to compact worker, but apply worker?

Comment (javaforfun) [inline src/raftstore/store/store.rs]: because we need change and save apply_state.truncated_state in apply worker

**Comment by siddontang** [comment]
Maybe we can remove `raftlog_gc.rs` and ignore the compact log handling when applying the CompactLog admin command.

**Comment by javaforfun** [comment]
If we need backward compatibility, the CompactLog admin command still need handling in apply worker. 
Because there may have some raft log not applied before restart,  we still have to handle it after restart, no matter we restart a single server or the whole cluster.

**Comment by ngaut** [comment]
We can ignore backward compatibility before 1.0 GA

**Comment by siddontang** [comment]
we can ignore the CompactLog and mark it as applied directly, later the GC scheduler will compact the log automatically.

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -1325,6 +1361,46 @@ impl Runner {
         }
     }
 
+    fn handle_local_compact_logs(&mut self, compacts: Vec<LocalCompactLog>) {
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: please add test

**Pull request closed without merging**